contributors/design-proposals/storage/container-storage-interface-skip-attach.md

```diff
@@ -0,0 +1,81 @@
+
+## Design
+### Volume plugins
+
+#### CSI volume plugin
+* Rework [`ProbeVolumePlugins`](https://github.com/kubernetes/kubernetes/blob/43f805b7bdda7a5b491d34611f85c249a63d7f97/pkg/volume/csi/csi_plugin.go#L58) to accept informer that watches CSIDriver. The plugin will store the informer for later.
+* Rework `Attach`, `Detach`, `VolumesAreAttached` and `WaitForAttach` to check for `CSIDriver` instance using the informer.
+	* If CSIDriver for the driver exists and it's attachable, perform usual logic.
+	* If CSIDriver for the driver exists and it's not attachable, return success immediately (basically NOOP).
+	* If CSIDriver for the driver does not exist, perform usual logic (i.e. treat the volume as attachable).
+	  * This keeps the behavior the same as in old Kubernetes version without CSIDriver object.
+	  * This also happens when CSIDriver informer has not been quick enough. It is suggested that CSIDriver instance is created **before** any pod that uses corresponding CSI driver can run.
+	    * In case that CSIDriver informer (or user) is too slow, CSI volume plugin `Attach()` will create `VolumeAttachment` instance and wait for (non-existing) external attacher to fulfill it. The CSI plugin shall recover when `CSIDriver` instance is created and skip attach. Any `VolumeAttachment` instance created here will be deleted on `Detach()`, see the next bullet.
+* In addition to the above, `Detach()` removes `VolumeAttachment` instance even if the volume is not attachable. This deletes `VolumeAttachment` instances created by old A/D controller or before `CSIDriver` instance was created.
+
+### A/D controller
+* A/D controller must pass a (shared) informer to watch CSIDriver and pass it to CSI volume plugin in [`ProbeVolumePlugins`](https://github.com/kubernetes/kubernetes/blob/8db5328c4c1f9467ab0d70ccb991a12d4675b6a7/cmd/kube-controller-manager/app/plugins.go#L82).
+
+### Kubelet / VolumeManager
+* Kubelet must create a (shared) informer to watch CSIDriver and pass it to CSI volume plugin in [`ProbeVolumePlugins`](https://github.com/kubernetes/kubernetes/blob/8db5328c4c1f9467ab0d70ccb991a12d4675b6a7/cmd/kubelet/app/plugins.go#L101).
+
+## API
+No API changes.
+
+## Upgrade
+This chapter covers:
+* Upgrade from old Kubernetes that has `CSISkipAttach` disabled to new Kubernetes with `CSISkipAttach` enabled.
+* Update from Kubernetes that has `CSISkipAttach` disabled to the same Kubernetes with `CSISkipAttach` enabled.
+* Creation of CSIDriver instance with non-attachable CSI driver.
+
+In all cases listed above, an "attachable" CSI driver becomes non-attachable. Upgrade does not affect attachable CSI drivers, both "old" and "new" Kubernetes processes them in the same way.
+
+For non-attachable volumes, if the volume was attached by "old" Kubernetes (or "new" Kubernetes before CSIDriver instance was created), it has `VolumeAttachment` instance. It will be deleted by `Detach()`, as it deletes `VolumeAttachment` instance also for non-attachable volumes.
```

Comment:
Is there a problem when controller-manager is newer but kubelet is older? The controller-manager will not create `VolumeAttachment` object but kubelet will be stuck waiting for `WaitForAttach` ?

Reply:
Admin should not enable the alpha feature in this case, I think. Can we solve it by documentation?

Reply:
What about for beta/GA?  The standard upgrade procedure is to update master first

Reply:
Or is the requirement to not update your drivers until both master+nodes are beyond some version?

Reply:
I think that whatever creates CSIDriver instances should create it only when both A/D controller and kubelet understand it. I.e. they're both at the right version and they have necessary alpha features enabled.

On downgrade, A/D controller gets updated first and it starts creating VolumeAttachment instances. It's up to the admin to start external attacher in this case, because CSI required it at that old version (or version with disabled alpha features).